### ISSUE TITLE
Simple instalation without cache

### ISSUE THREAD

**Issue by vanheck** [python-poetry/poetry#3374]
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] I have searched the [documentation](https://python-poetry.org/docs/) and believe that my question is not covered.

### TL;DR
I'd like to use `pyproject.toml` for instaling dependencies for project with minimal impact on size where the docker container provides sufficient isolation. E.G.:
```
poetry install --no-root --no-cache --no-interaction
```

### Details, blockers and other similar issues
I'd like to use poetry in docker as simple as `pip install --no-cache-dir -r requirements.txt` (with cleared cache), but have some trouble. I have a local dependency which is wanted to be editable in the same local folder. 
```pyproject.toml
...
[tool.poetry.dependencies]
python = "^3.8"
numpy = "^1.19.4"
pandas = "^1.1.4"
jupyterlab = "^2.2.9"

[tool.poetry.dev-dependencies]
# is installed as "python setup.py develop" and can be edited in mypackage folder
mypackage = {path = "./mypackage", develop = true}
```

Installation in docker:
```Dockerfile
FROM python:3.8-slim-buster

WORKDIR /lab

RUN pip install poetry --no-cache-dir && \
    poetry config virtualenvs.create false
COPY poetry.lock pyproject.toml .

# my local dependency
COPY ./mypackage ./mypackage

# too big
RUN poetry install --no-root --no-interaction

EXPOSE 9180
CMD ["jupyter", "lab", "--allow-root", "--ip=0.0.0.0", "--port=9180", "."]
```
I've tried also `RUN poetry install --no-root --no-interaction && poetry cache clear --no-interaction --all pypi`, but size is the same. 

I found https://github.com/python-poetry/poetry/issues/1879#issuecomment-592133519, but there is too many rows and dockerfile is being complex.

I've also tried https://github.com/python-poetry/poetry/issues/1879#issuecomment-650384931:
```
RUN poetry export -f requirements.txt --dev --without-hashes | pip install --no-cache-dir -r /dev/stdin
```
This is fine workaround but its not suitable for `mypackage` dependency. Export row of package is `mypackage @ /home/vh/test-docker/mypackage; python_version >= "3.8" and python_version < "4.0"` and pip searching this package on pypi and fail.

When I remove the dependency and run previous command without `--dev` param:
```
RUN poetry export -f requirements.txt -v --without-hashes | pip install --no-cache-dir -r /dev/stdin && \
         cd mypackage && \
         poetry install --develop mypackage && \
         cd ..
```
Also crashes with error that `--develop` param was removed with release poetry v1.0.0. But I found opened issue: https://github.com/python-poetry/poetry/issues/2572.

**Comment by sinoroc** [comment]
Personal opinion:

I see how it might be a helpful feature. On the other hand I can't help but think that it should be quite easy to just delete the cache and other build artefacts and what not right after the installation:
```
RUN poetry install ... && rm -rf ~/.cache/pypoetry/{cache,artifacts}
```
We might spare us adding complexity into _poetry_'s code base. But maybe also it is not as simple as that and a `--no-cache` feature makes sense.

**Comment by vanheck** [comment]
Yes, these dirs contains some instalation balast. I removed them and container is much smaller, it works thanks.

**Comment by max-sixty** [comment]
Thanks @sinoroc . For others who come across this, I made a slight change:

With a dockerfile that's like:

```sh
RUN poetry config virtualenvs.create false \
    && poetry config experimental.new-installer false \
    && poetry install --no-root --no-interaction --no-ansi 
```

The cache went to `~/.cache/pip`, so I changed the command to

```
&& rm -rf ~/.cache/{pip,pypoetry}
```

---

Instead of adding the complexity into poetry, how reasonable would it be to have an option to pass flags along to pip? That way there's an escape hatch for all options without having to build a full facade?

**Comment by cbrochtrup** [comment]
Another option to keep the cache out of the final image is to use [docker BuildKit's cache feature](https://docs.docker.com/develop/develop-images/build_enhancements/). I've used it with success.

Example

```Dockerfile
# syntax=docker/dockerfile:1.2

COPY poetry.lock /
RUN --mount=type=cache,target=/home/.cache/pypoetry/cache \
    --mount=type=cache,target=/home/.cache/pypoetry/,artifacts \
    poetry install
```

This way, docker keeps your poetry cache around for quick builds, layer sharing, etc. but not in the final image.

**Comment by JohnPreston** [comment]
I have a non 1.2 format way to do this in here: https://github.com/compose-x/ecs_composex/blob/main/cli.Dockerfile
Using an intermediary container, I build the .whl and give that to my final container, which saves me from installing pip and everything the project does not need, and at the same time allows me to verify that people installs from pip will work all the same.
Downside is you are "cacing" the .whl in the image as RUN and COPY create two different layers but unless your py code is in MBs, that should really not be a problem.

**Comment by rc-mattschwager** [comment]
It looks like `--no-cache` only removes the `~/.cache/pypoetry/cache/` directory and not `~/.cache/pypoetry/artifacts/`.

I tested with the following:

```
$ cat Dockerfile 
FROM python:3.10-slim

WORKDIR /app
COPY . ./

RUN python -m pip install poetry==1.2.2
RUN python -m poetry install

CMD ["/bin/bash"]
```

```
$ cat pyproject.toml 
[tool.poetry]
name = "poetry-test"
version = "0.1.0"
description = ""
authors = ["Your Name <you@example.com>"]
readme = "README.md"

[tool.poetry.dependencies]
python = "^3.10"
httpx = "^0.23.0"


[build-system]
requires = ["poetry-core"]
build-backend = "poetry.core.masonry.api"
```

```
$ docker build -t poetry-test .
$ docker run -it poetry-test 
root@f47a0c787efb:/app# ls ~/.cache/pypoetry/
artifacts  cache  virtualenvs
```

When I add `--no-cache` I get the following:

```
$ docker run -it poetry-test 
root@a55245819f52:/app# ls ~/.cache/pypoetry/
artifacts  virtualenvs
```

The `artifacts` directory is still present. This directory is often quite large in my builds because it contains all the `whl` files from installation. Should `poetry` not include this directory either when `--no-cache` is present?

**Comment by TBBle** [comment]
@rc-mattschwager: I think you've hit #6445; `--no-cache` is only for one kind of cache right now, and it's not documented clearly what that is.

There doesn't appear to be anything in Poetry that will auto-nuke the "artifacts" cache at the end of the process, but perhaps
```
RUN python -m poetry install && poetry cache clear pypi --all
```
will do what you want?

For your test-file, you might also want to pass `--no-cache` to the `pip` command; based on #6445 that will do what you expected for that invocation, so you don't end up with cached Poetry-and-dependency wheels in that container image layer.

**Comment by rc-mattschwager** [comment]
Ahh, I see, I will follow along in #6445. Thanks for pointing me in the right direction!

Based on my testing, `cache clear` doesn't remove the `artifacts` directory. Perhaps there's no first-class method for doing that, and I'll have to do a `rm` or a multi-stage build. Although, like #6445, I agree it would be nice to have a flag similar to `pip`'s `--no-cache-dir` in `poetry`.
